### PR TITLE
Fix usage fault in flash shell sample for RT1064 EVK

### LINKED ISSUE #44043: Usage fault when running flash shell sample on RT1064 EVK
**Describe the bug**
The flash shell sample immediately encounters a usage fault on the RT1064 EVK.

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=mimxrt1064_evk samples/drivers/flash/flash_shell
3. make

**Expected behavior**
The sample should start correctly, and allow the user to use the flash device

**Logs and console output**
```
Found flash controller FLEXSPI1.
Flash I/O commands can be run.

[00:00:00.011,962] <err> os: ***** USAGE FAULT *****
[00:00:00.011,993] <err> os:   Illegal use of the EPSR
[00:00:00.012,023] <err> os: r0/a1:  0x00000000  r1/a2:  0x7000e0b0  r2/a3:  0x00000003
[00:00:00.012,023] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x70007987
[00:00:00.012,023] <err> os:  xpsr:  0x20000024
[00:00:00.012,023] <err> os: Faulting instruction address (r15/pc): 0x00000000
[00:00:00.012,084] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:00.012,084] <err> os: Fault during interrupt handling

[00:00:00.012,115] <err> os: Current thread: 0x80000638 (shell_uart)
[00:00:00.101,226] <err> os: Halting system
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.13.2
 - Sha 452133cf

**Additional context**
This is caused by a RWW hazard, and was introduced with 5af7fe66. The FlexSPI driver should never access flash while attempting to write data to the FlexSPI device.

### PR DESCRIPTION
**Pull request by danieldegrasse**
Fix usage fault in flash shell sample for RT1064 EVK. This fault is cause by a read-while-write hazard, the FlexSPI driver cannot access flash memory while it is actively writing data to the flash device. This PR reverts the changes introduced in #42230, which moved device objects from RAM into flash, triggered a RWW bug, and additionally moves the memc driver data into ram to prevent future RWW hazards. Comments have also been added around critical sections to warn developers that flash access at these times will cause RWW bugs.

Fixes #44043

### CODE REVIEW

File: drivers/flash/flash_mcux_flexspi_hyperflash.c [hunk 1]
```diff
@@ -657,10 +680,15 @@ static const struct flash_driver_api flash_flexspi_hyperflash_api = {
 		},							\
 	};								\
 									\
+	struct flash_flexspi_hyperflash_data				\
```
Comment (pdgendt) [inline drivers/flash/flash_mcux_flexspi_hyperflash.c:683]: Add `static`

File: drivers/flash/flash_mcux_flexspi_mx25um51345g.c [hunk 2]
```diff
@@ -535,10 +561,15 @@ static const struct flash_driver_api flash_flexspi_nor_api = {
 		},							\
 	};								\
 									\
+	struct flash_flexspi_nor_data					\
```
Comment (pdgendt) [inline drivers/flash/flash_mcux_flexspi_mx25um51345g.c:564]: Same: Add `static`

File: drivers/flash/flash_mcux_flexspi_nor.c [hunk 3]
```diff
@@ -547,12 +573,16 @@ static const struct flash_driver_api flash_flexspi_nor_api = {
 			.write_block_size = NOR_WRITE_SIZE,		\
 			.erase_value = NOR_ERASE_VALUE,			\
 		},							\
+	};								\
+	struct flash_flexspi_nor_data					\
```
Comment (pdgendt) [inline drivers/flash/flash_mcux_flexspi_nor.c:577]: Same: Add `static`

File: drivers/memc/memc_mcux_flexspi_hyperram.c [hunk 4]
```diff
@@ -175,15 +182,19 @@ static int memc_flexspi_hyperram_init(const struct device *dev)
 #define MEMC_FLEXSPI_HYPERRAM(n)				  \
 	static const struct memc_flexspi_hyperram_config	  \
 		memc_flexspi_hyperram_config_##n = {		  \
-		.controller = DEVICE_DT_GET(DT_INST_BUS(n)),	  \
 		.port = DT_INST_REG_ADDR(n),			  \
 		.config = MEMC_FLEXSPI_DEVICE_CONFIG(n),	  \
 	};							  \
 								  \
+	struct memc_flexspi_hyperram_data			  \
```
Comment (pdgendt) [inline drivers/memc/memc_mcux_flexspi_hyperram.c:189]: Same: Add `static`

**Review by pdgendt — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_mcux_flexspi_mx25um51345g.c (hunk 2 above)
Comment (danieldegrasse) [reply drivers/flash/flash_mcux_flexspi_mx25um51345g.c]: fixed in latest push

File: drivers/flash/flash_mcux_flexspi_nor.c (hunk 3 above)
Comment (danieldegrasse) [reply drivers/flash/flash_mcux_flexspi_nor.c]: fixed in latest push

**Review by pdgendt — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Pull request merged**